docs/source/nested.rst

```diff
@@ -39,6 +39,20 @@ nested_tensor([
   tensor([3., 4., 5., 6., 7.], device='cuda:0')
 ])
 
+Note that the passed Tensors are being copied into a contiguous piece of memory. The resulting
+NestedTensor allocates new memory to store them and does not keep a reference.
+
+In order to form a valid NestedTensor the passed Tensors also all need to match in dimension.
+
+At this moment we only support one level of nesting, i.e. a simple, flat list of Tensors. In the future
```

Comment:
flat list of tensors could very well be a flat list of tensors with two unmatched dims (like images with different resolutions)

Reply:
So, aside from the dimension, semantically none of the other attributes need to match (e.g. dtype, device, numel, sizes, strides, etc.). In practice [the constructor](https://github.com/pytorch/pytorch/blob/09f79e94acce30e2002063322c9fe50d116361f1/aten/src/ATen/native/nested/NestedTensorMath.cpp#L267-L307) flattens each input, concatenates them all together and records their original sizes. This concatenated piece of memory and the sizes form the NestedTensor. For the concatenation operation to succeed the dtype and device of the input Tensors need to match, but it'd be easy to extend the constructor to detect when this isn't the case and use either the default dtype / device or require the user to pass one.

Reply:
ah, I see, so many dims can match. am I right that the nested tensor always "batches" on first dimension (like torch.stack with default arg would do)? (it could have been last one or some arbitrary one in theory)

Reply:
Yes, it's always the first dimension. A future, future extension of torch.stack would automatically create a NestedTensor if passed variably sized Tensors and thus could support arbitrary dimensions. Do you have a use case in mind? I suppose we could add a `dim` argument to the `nested_tensor` constructor.